mitiq/folding_cirq.py

```diff
@@ -1,76 +1,197 @@
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number >= 1."""
+    Note:
+        Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
+    """
+    folded = deepcopy(circuit)
 
-    if not (stretch >= 1):
-        raise ValueError("The stretch factor must be a real number >= 1.")
+    if np.isclose(stretch, 1.0, atol=1e-2):
+        return folded
 
-    if stretch <= 3:
-        return stretcher(circuit, stretch)
-    else:
-        # recursive iterations for stretch > 3
-        _ = local_folding(circuit, 3, stretcher)
-        return local_folding(_, stretch / 3, stretcher)
+    if not 1 < stretch <= 3:
+        raise ValueError("The stretch factor must be a real number between 1 and 3.")
 
+    if "seed" in kwargs.keys():
+        np.random.seed(kwargs.get("seed"))
 
-# ==================================================
-# Circuit level folding
-# ==================================================
+    ngates = len(list(folded.all_operations()))
+    num_to_fold = int(ngates * (stretch - 1.0) / 2.0)
 
+    for _ in range(num_to_fold):
+        # TODO: This allows for gates to be folded more than once, and folded gates to be folded.
+        #  Should this be allowed?
+        moment_index = np.random.choice(len(folded))
+        gate_index = np.random.choice(len(folded[moment_index]))
+        # TODO: It could be expensive to call fold_gate...(...) which makes a deepcopy of the circuit each call.
+        #  ==> Possible fix: Have fold_gate_at_index_in_moment(...) modify the circuit in place.
+        folded = fold_gate_at_index_in_moment(folded, moment_index, gate_index)
 
+    return folded
+
+
+def fold_local(circuit: Circuit, stretch: float, fold_method: Callable = fold_gates_from_left, **kwargs) -> Circuit:
+    """Returns a folded circuit by folding gates according to the input fold method.
+
+    Args:
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by.
+        fold_method: Function which defines the method for folding gates.
+                    (e.g., Randomly selects gates to fold, folds gates starting from left of circuit, etc.)
+
+                    Must have signature
+
+                    def fold_method(circuit: Circuit, stretch: float, **kwargs):
+                        ...
+
+                    and return a circuit.
+    """
+    folded = deepcopy(circuit)
+
+    if np.isclose(stretch, 1.0, atol=1e-2):
+        return folded
+
+    if not 1 < stretch <= MAX_STRETCH_FACTOR:
+        raise ValueError(f"The stretch factor must be a real number between 1 and {MAX_STRETCH_FACTOR}.")
+
+    while stretch > 1.:
+        this_stretch = 3. if stretch > 3. else stretch
+        # TODO: This also allows folding gates that have already been folded. Should this be allowed?
+        folded = fold_method(folded, this_stretch, **kwargs)
```

Comment:
Good point. As before I don't think it should be. The change, I think, would be to add a `circ_copy = circuit` outside the loop and then do
```suggestion
        folded = fold_method(circuit, this_stretch, **kwargs)
```